File: Cpp/AdventOfCode/src/2015/d14_ReindeerRaces.cpp
#pragma once

#include "2015/d14_ReindeerRaces.h"

SOLUTION(2015, 14) {
    //Rudolph can fly 22 km/s for 8 seconds, but then must rest for 165 seconds.
    struct Reindeer
    {
        u32 Speed{ 0 };
        u32 RunTime{ 0 };
        u32 RestTime{ 0 };
    };

    constexpr Reindeer ParseRacer(std::string_view line) {
        auto s = Constexpr::Split(line, " ");
        Reindeer result;
        Constexpr::ParseNumber(s[3], result.Speed);
        Constexpr::ParseNumber(s[6], result.RunTime);
        Constexpr::ParseNumber(s[13], result.RestTime);

        return result;
    }

    constexpr u32 FindDistance(Reindeer who, u32 seconds) {
        u32 cycleTime = who.RunTime + who.RestTime;
        u32 cycles = seconds / cycleTime;
        u32 remainder = seconds % cycleTime;

        u32 distance = who.Speed * cycles * who.RunTime;

        if (remainder <= who.RunTime) {
            distance += who.Speed * remainder;
        }
        else {
            distance += who.Speed * who.RunTime;
        }

        return distance;
    }


    constexpr std::vector<u32> GetPoints(const std::vector<Reindeer>& racers, u32 seconds) {
        auto points = std::vector<u32>(racers.size(), 0);
        auto distances = std::vector<u32>(racers.size(), 0);
        for (u32 second = 1; second <= seconds; second++) {
            u32 best = 0;
            for (size_t index = 0; index < racers.size(); index++) {
                distances[index] = FindDistance(racers[index], second);
                best = std::max(best, distances[index]);
            }
            for (size_t index = 0; index < racers.size(); index++) {
                points[index] += distances[index] == best;
            }
        }

        return points;
    }

    PART_ONE() {
        //One liner, just for fun. :)
        return Constexpr::ToString(RunAllReturnMax(ParseLines(Lines, ParseRacer), FindDistance, 2503));
        /*
        auto racers = ParseLines<Reindeer>(lines, ParseRacer);
        u32 best = 0;
        for (const auto& racer : racers) {
            best = std::max(best, FindDistance(racer, 2503));
        }

        return Constexpr::ToString(best);
        */
    }
    PART_TWO() {
        auto racers = ParseLines(Lines, ParseRacer);
        auto points = GetPoints(racers, 2503);
        
        return Constexpr::ToString(*std::max_element(points.cbegin(), points.cend()));
    }

    TESTS() {
        static_assert(FindDistance({ 14,10,127 }, 10) == 140);
        static_assert(FindDistance({ 14,10,127 }, 11) == 140);
        static_assert(FindDistance({ 14,10,127 }, 1000) == 1120);

        static_assert(FindDistance({ 16,11,162 }, 10) == 160);
        static_assert(FindDistance({ 16,11,162 }, 11) == 176);
        static_assert(FindDistance({ 16,11,162 }, 1000) == 1056);

        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d14_reindeer_races.rs
pub fn part_one(lines: &Vec<String>) -> String {
    lines.iter().map(|l| Reindeer::parse(l)).map(|r| find_distance(&r, 2503)).max().unwrap().to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    let racers = lines.iter().map(|l| Reindeer::parse(l)).collect();
    get_points(&racers, 2503).to_string()
}

struct Reindeer {
    speed: usize,
    run_time: usize,
    rest_time: usize
}

impl Reindeer {
    fn parse(line: &str) -> Self {
        let s: Vec<&str> = line.split(" ").collect();
        Self {
            speed: s[3].parse().unwrap(),
            run_time: s[6].parse().unwrap(),
            rest_time: s[13].parse().unwrap()
        }
    }
}

fn find_distance(who: &Reindeer, seconds: usize) -> usize {
    let cycle_time = who.run_time + who.rest_time;
    let cycles = seconds / cycle_time;
    let remainder = seconds % cycle_time;
    let mut distance = who.speed * cycles * who.run_time;

    if remainder <= who.run_time {
        distance += who.speed * remainder;
    } else {
        distance += who.speed * who.run_time;
    }

    distance
}

fn get_points(racers: &Vec<Reindeer>, seconds: usize) -> usize {
    let mut bests = vec![0; seconds + 1];
    for second in 1..=seconds {
        bests[second] = racers.iter().map(|r| find_distance(r, second)).max().unwrap();
    }

    racers.iter().map(|r| (1usize..=seconds)
        .map(|second| {
            if find_distance(r, second) == bests[second] {
                1usize
            } else {
                0usize
            }
        }
    ).sum::<usize>()
    ).max().unwrap()
}